_posts/2020-06-01-totw-116.md

```diff
@@ -0,0 +1,162 @@
+ public:
+  explicit Foo(const std::string& content) : content_(content) {}
+  const std::string& content() const { return content_; }
+
+ private:
+  const std::string& content_;
+};
+</pre>
+
+It looks reasonable. But what happens if we build a `Foo` from a string literal?
+
+<pre class="prettyprint lang-cpp code">
+void Func() {
+  Foo foo("something");
+  LOG(INFO) &lt;&lt; foo.content();  // BOOM!
+}
+</pre>
+
+When creating `foo`, the `content_` member gets bound to the temporary
+`std::string` object that was created from the literal and passed to the
+constructor. The temporary string goes out of scope at the end of the line where
+it was created. Now `foo.content_` is a reference to an object that no longer
+exists. Accessing it is undefined behavior and anything can happen, from working
+fine in tests to going really wrong in production.
+
+## A Solution: Use Pointers
+
+In our example, the simplest solution is probably to pass and store the string
+by value. But let's assume that we need to refer to the original argument, e.g.
+because it's not a string, but some more interesting type. The solution is to
+pass the argument by pointer:
+
+<pre class="prettyprint lang-cpp code">
+class Foo {
+ public:
+  // Do not forget this comment:
+  // Does not take ownership of content, which must refer to a valid string that
+  // outlives this object.
+  explicit Foo(const std::string* content) : content_(content) {}
+  const std::string& content() const { return *content_; }
+
+ private:
+  const std::string* const content_;  // not owned, can't be null
+};
+</pre>
+
+Now the following will simply fail to compile:
+
+<pre class="prettyprint lang-cpp code">
+std::string GetString();
+void Func() {
+  Foo foo1(&GetString());  // error: taking the address of a temporary of
+                           // type 'std::string'
+  Foo foo2(&"something");  // error: no matching constructor for initialization
+                           // of 'Foo'
+}
+</pre>
+
+And it will be pretty clear at call site that the object might keep the address
+of the argument:
+
+<pre class="prettyprint lang-cpp code">
+void Func2() {
+  std::string content = GetString();
+  Foo foo(&content);
+}
+</pre>
+
+## One Step Further, One Less Comment: Storing a Reference
+
+You might have noticed that we say twice that the pointer cannot be null and
+that it is not owned, once in the documentation of the constructor then in the
+comment for the instance variable. Is this necessary? Consider this:
+
+<pre class="prettyprint lang-cpp code">
+class Baz {
+ public:
+  // Does not take any ownership, and all pointers must refer to valid objects
+  // that outlive the one constructed.
+  explicit Baz(const Arg1* arg1, Arg2* arg2) : arg1_(*arg1), arg2_(*arg2) {}
+
+ private:
+  // It is now clear that we do not have ownership and that the references can't
+  // be null.
+  const Arg1& arg1_;
+  Arg2& arg2_;  // Yes, non-const references are style-compliant! [Tip #87](/tips/87)
```

Comment:
We haven't published 87.

Reply:
Should be removed in the most recent commit, but I still don't totally know how github PRs work :P.